### PR TITLE
Add popups component: tooltips and infoboxes

### PR DESCRIPTION
**Pull request by kandre**

### CODE REVIEW

**Comment by kandre** [comment]
@oterral we have two ajax request (identify and htmlPopup) within loops, what does not allow fixmyjs module ('don't make functions within a loop error'). Could we bypass it ?

**Comment by kandre** [comment]
It works now, thanks

**Comment by kandre** [comment]
[Testlink](http://mf-geoadmin3.dev.bgdi.ch/emapis-kan/#)

**Comment by kandre** [comment]
It is ready for review

File: index.html [hunk 1]
```diff
@@ -46,6 +46,37 @@
...
+      </div>
+      <div class="tooltip-title"></div>
+    </div>
+
+    <div id="reducedTooltip">
+      <div class="popup-buttons">
+        <button id="tooltipBtnCloseReduced" class="close" type="button"><span>&times;</span></button>
+      </div>
+      <div class="tooltip-title tooltip-title-reduced"></div>
+    </div>
```
Comment (oterral) [inline index.html]: Do you really need a reducedTooltip?  Can't you just add a css like in map.geo.admin.ch?

File: index.html [hunk 2]
```diff
@@ -70,9 +101,9 @@ <h4 id="modalLabel">Infobox</h4>
           <div class="modal-body"></div>
         </div>
       </div>
-    </div>
+    </div> -->
```
Comment (oterral) [inline index.html]: remove the commented html

File: js/popup.js [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+  var tooltipBtnClose = $('#tooltipBtnClose');
+  var tooltipBtnReduce = $('#tooltipBtnReduce');
+  var reducedTooltip = $('#reducedTooltip');
+  var tooltipBtnCloseReduced = $('#tooltipBtnCloseReduced');
+  var infoboxBtnClose = $('#infoboxBtnClose');
+  var infoboxBtnReduce = $('#infoboxBtnReduce');
+  var reducedInfobox = $('#reducedInfobox');
+  var infoboxBtnCloseReduced = $('#infoboxBtnCloseReduced');
+  var parser = new ol.format.GeoJSON();
+  var htmlUrl = 'http://mf-chsdi3.dev.bgdi.ch/kan_chsdi_emapis_entwaesserung/rest/services/{Topic}/MapServer/{Layer}/{Feature}/htmlPopup'; // adapt to prod url when needed
```
Comment (oterral) [inline js/popup.js]: Not very generic

File: index.html [hunk 4]
```diff
@@ -70,9 +101,9 @@ <h4 id="modalLabel">Infobox</h4>
...
     <script src="app.js"></script>
-    <script src="http://api3.geo.admin.ch/loader.js?lang=de"></script>
+    <script src="http://api3.geo.admin.ch/loader.js?mode=debug"></script>
```
Comment (kandre) [inline index.html]: I will change debug mode

File: js/popup.js (hunk 3 above)
Comment (kandre) [inline js/popup.js]: Yes, it is the branch where @AFoletti works

File: js/popup.js [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+              '</div>' +
+            '</div>';
+          feature.set('htmlpopup', htmlpopup);
+          showFeatures(layerToQuery.getExtent(), size,
+              [feature]);
+        }
+      } else { // queryable bod layers
+        var idWmsLayer = 'all:' + layerToQuery.get('id');
+        var idApiLayer = 'all:' + layerToQuery.id;
+        var layerParam = layerToQuery.get('id') ? idWmsLayer : idApiLayer;
```
Comment (oterral) [inline js/popup.js]: all-in-one : `layerParam = 'all:' + (layerToQuery.get('id') || layerToQuery.id)`

File: js/popup.js [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+  var onHtmlPopupSuccess = function(html) {
+    showTooltip(html);
+  };
+
+  // Show and highlight the found features found
+  var showFeatures = function(mapExtent, size, foundFeatures) {
+    if (foundFeatures && foundFeatures.length > 0) {
+
+      for (var i = 0; i < foundFeatures.length; i++) {
+        value = foundFeatures[i];
```
Comment (oterral) [inline js/popup.js]: `var value = `

File: js/popup.js [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+                        .replace('{Topic}', 'all') // might be nicer for emapis
+                        .replace('{Layer}', value.layerBodId)
+                        .replace('{Feature}', value.featureId);
+
+          // Ajax html popup request
+          $.ajax({
+            type: 'GET',
+            url: htmlUrl,
+            data: {
+              lang: 'de',
```
Comment (oterral) [inline js/popup.js]: make it language dependent

File: js/popup.js [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+    } else {
+      reducedInfobox.offset({left:0});
+    }
+    if (reducedInfobox.is(':visible')) {
+      offset = reducedInfobox.offset().left + reducedInfobox.width();
+      reducedTooltip.offset({left:offset});
+    } else {
+      reducedTooltip.offset({left:0});
+    }
+  };
```
Comment (oterral) [inline js/popup.js]: if it's just a problem of  style you can use `float: left` when the popup is reduced no nee to calculate the left position

**Comment by kandre** [comment]
I made the corrections, including simplyfying jquery clicks event by using more this.parentElement method

File: less/popup.less [hunk 9]
```diff
@@ -0,0 +1,149 @@
+// Styling popups
+
+#tooltip, #infobox {
```
Comment (oterral) [inline less/popup.less]: use a common css class pls

File: less/popup.less [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+#tooltip {
+  max-width: 450px;
+}
+
+#infobox {
+  max-width: 620px;
+  max-heigth: 420px;
+}
+
+.tooltip-title, .infobox-title {
```
Comment (oterral) [inline less/popup.less]: here too a common css class is nicer

File: less/popup.less [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  font-size: 11px!important;
+}
+
+.popup-buttons {
+  top: 6px;
+  position: absolute;
+  right: 10px;
+}
+
+button.close {
```
Comment (oterral) [inline less/popup.less]: Use `.btn-close`

File: less/popup.less [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+  font-size: 26px;
+  font-weight: bold;
+  padding-left: 7px 
+}
+
+button.reduce {
```
Comment (oterral) [inline less/popup.less]: Ude `.btn-reduce`

File: less/popup.less [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+  min-height: 40px!important;
+  width: 190px!important;
+  min-width: 190px!important;
+  bottom: 26px;
+  left: 0px!important;
+  overflow: hidden; // should not only be applied on reduce !
+  top: auto!important;
+  bottom: 66px;
+
+  .tooltip-title, .infobox-title {
```
Comment (oterral) [inline less/popup.less]: Use a common css class

File: less/popup.less [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+  background-color: transparent;
+  border-collapse: collapse;
+  border-spacing: 0;
+}
+
+tbody {
+  display: table-row-group;
+  vertical-align: middle;
+  border-color: inherit;
+}
```
Comment (oterral) [inline less/popup.less]: Can you use a more less syntax to separate what is needed only for infobox and what is need only for tooltip

File: js/popup.js [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+
+    // Append html 
+    $('.legend-container').remove();
+    infobox.append(html);
+
+    // show tooltip
+    infobox.show();
+
+    // Make it draggable
+    draggable(infobox);
```
Comment (oterral) [inline js/popup.js]: The draggable event should be added only once, not each time you open the dialog box.

**Comment by kandre** [comment]
Can we get rid of closeand reducecss classes ? we have btn-close and btn-reduce now

=> I use close class for the reduce button to use the bootstrap styles for minus icon like on close icon

File: js/popup.js [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+  $('body').on('click', '.popup-reduce', function(event) {
+    $(this).removeClass('popup-reduce');
+    event.stopPropagation();
+  });
+
+  // Event click on close button on infobox
+  $('.infobox-btn-close').click(function(event) {
+    infobox.hide();
+    event.stopPropagation();
+  });
```
Comment (oterral) [inline js/popup.js]: still needed?

File: js/popup.js [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+  };
+
+  // Find features for all type of layers
+  var findFeatures = function(coordinate, size, mapExtent) {
+    var identifyUrl = 'http://mf-chsdi3.dev.bgdi.ch/kan_chsdi_emapis_entwaesserung/rest/services/all/MapServer/identify'; // adapt to prod url when needed
+    var layersToQuery = getLayersToQuery();
+    var pixel = map.getPixelFromCoordinate(coordinate);
+    for (var i = 0; i < layersToQuery.length; i++) {
+      var layerToQuery = layersToQuery[i];
+      if (isVectorLayer(layerToQuery)) {
```
Comment (oterral) [inline js/popup.js]: we don't have vector features I think you can remove this part

File: less/popup.less [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+#tooltip {
+  max-width: 450px;
+}
+
+#infobox {
+  max-width: 620px;
+  max-heigth: 420px;
+
```
Comment (oterral) [inline less/popup.less]: a `{` is missing

File: js/translation.js [hunk 19]
```diff
@@ -6,10 +6,16 @@ var addTranslation = function(objPermalink, translator) {
...
+
+    // Measure tool translation
     var measure = $('#measure');
     measure.text(translator.get(measure.attr('data-value')));
     var helpMsg = $('.tooltip-help').attr('data-value');
     $('.tooltip-help').text(translator.get(helpMsg));
+    
+    // Popups (tooltip and infobox
+    $('.tooltip-title').text(translator.get('tooltipTitle'));
+    $('.infobox-title').text(translator.get('infoboxTitle'));
```
Comment (oterral) [inline js/translation.js]: can you use a better selector `#tooltip .popup-title` pls? with that you can remove useless css class

**Comment by oterral** [comment]
Ok code looks good apart some minor things.
I've found few things to modify:
  -  Original position of infobox should be on th left, close to the catalog
  -  Original position of tooltip should be on the right, close to navigation buttons
  -  Max-height of tooltip popup is too big, should be the same as tooltip popup
  -  Re-initailize the position of popups when you reopen it  
  -  Reload popup content  on language change

**Comment by kandre** [comment]
Max-height of tooltip popup is too big, should be the same as tooltip popup:

@oterral what do you meant, it is the same as itself ;) ?

**Comment by oterral** [comment]
no, tooltip is too big

**Comment by kandre** [comment]
Ok I see the point, with adding a missing overflow-y: scroll for the tooltip

**Comment by oterral** [comment]
Thks for this!!!

**Pull request merged**